presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java

```diff
@@ -288,25 +282,107 @@ public void setWindowFunctions(QuerySpecification node, List<FunctionCall> funct
         return windowFunctions.get(query);
     }
 
-    public void setOutputDescriptor(RelationType descriptor)
+    public Scope getScope(Node node)
     {
-        outputDescriptor = descriptor;
+        return tryGetScope(node).orElseThrow(() -> new IllegalArgumentException(String.format("Analysis does not contain information for node: %s", node)));
     }
 
-    public RelationType getOutputDescriptor()
+    public Optional<Scope> tryGetScope(Node node)
+    {
+        // TODO check performance impact for planning of getScope method
+        if (scopes.containsKey(node)) {
+            return Optional.of(scopes.get(node));
+        }
+
+        Optional<Node> root = roots.stream()
+                .filter(i -> nodeContains(i, node))
+                .findAny();
+
+        if (root.isPresent()) {
+            GetScopeVisitor visitor = new GetScopeVisitor(scopes, node);
+            visitor.process(root.get(), new LinkedList<>());
+            return visitor.getResult();
+        }
+        return Optional.empty();
+    }
+
+    public Node getRoot()
     {
-        return outputDescriptor;
+        return roots.get(0);
     }
 
-    public void setOutputDescriptor(Node node, RelationType descriptor)
+    public Scope getRootScope()
+    {
+        return getScope(getRoot());
+    }
+
+    private static class GetScopeVisitor
```

Comment:
One thing that bugs me about this is that it assumes there's a link between lexical structure, AST structure and scoping rules. One case where this doesn't hold is lateral joins, for instance (per my other comment). I can imagine other scenarios in the future where this may not be true, such as a sequences of statements, procedures, etc.